packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -20,64 +23,418 @@ const double _kHandleSize = 22.0;
+const double _kToolbarContentDistanceBelow = _kHandleSize - 2.0;
 const double _kToolbarContentDistance = 8.0;
 
 /// Manages a copy/paste text selection toolbar.
-class _TextSelectionToolbar extends StatelessWidget {
+class _TextSelectionToolbar extends StatefulWidget {
   const _TextSelectionToolbar({
     Key key,
     this.handleCut,
     this.handleCopy,
     this.handlePaste,
     this.handleSelectAll,
+    this.isAbove,
   }) : super(key: key);
 
   final VoidCallback handleCut;
   final VoidCallback handleCopy;
   final VoidCallback handlePaste;
   final VoidCallback handleSelectAll;
 
+  // When true, the toolbar fits above its anchor and will be positioned there.
+  final bool isAbove;
+
+  @override
+  _TextSelectionToolbarState createState() => _TextSelectionToolbarState();
+}
+
+class _TextSelectionToolbarState extends State<_TextSelectionToolbar> with TickerProviderStateMixin {
+  // Whether or not the overflow menu is open. When it is closed, the menu
+  // items that don't overflow are shown. When it is open, only the overflowing
+  // menu items are shown.
+  bool _overflowOpen = false;
+
+  FlatButton _getItem(VoidCallback onPressed, String label) {
+    assert(onPressed != null);
+    return FlatButton(
+      child: Text(label),
+      onPressed: () {
+        setState(() {
+          _overflowOpen = false;
+        });
+        onPressed();
+      },
+    );
+  }
+
   @override
   Widget build(BuildContext context) {
     final MaterialLocalizations localizations = MaterialLocalizations.of(context);
     final List<Widget> items = <Widget>[
-      if (handleCut != null) FlatButton(child: Text(localizations.cutButtonLabel), onPressed: handleCut),
-      if (handleCopy != null) FlatButton(child: Text(localizations.copyButtonLabel), onPressed: handleCopy),
-      if (handlePaste != null) FlatButton(child: Text(localizations.pasteButtonLabel), onPressed: handlePaste),
-      if (handleSelectAll != null) FlatButton(child: Text(localizations.selectAllButtonLabel), onPressed: handleSelectAll),
+      if (widget.handleCut != null)
+        _getItem(widget.handleCut, localizations.cutButtonLabel),
+      if (widget.handleCopy != null)
+        _getItem(widget.handleCopy, localizations.copyButtonLabel),
+      if (widget.handlePaste != null)
+        _getItem(widget.handlePaste, localizations.pasteButtonLabel),
+      if (widget.handleSelectAll != null)
+        _getItem(widget.handleSelectAll, localizations.selectAllButtonLabel),
     ];
 
     // If there is no option available, build an empty widget.
     if (items.isEmpty) {
       return Container(width: 0.0, height: 0.0);
     }
 
-    return Material(
-      elevation: 1.0,
-      child: Container(
-        height: _kToolbarHeight,
-        child: Row(mainAxisSize: MainAxisSize.min, children: items),
+    final Align child = Align(
+      alignment: Alignment.topRight,
+      heightFactor: 1.0,
+      widthFactor: 1.0,
+      child: AnimatedSize(
```

Comment:
What are the incoming constraints for AnimatedSize? My guess is that they end up to be tight giving AnimatedSize no choice in what size it can be?

Reply:
Unfortunately I couldn't check the constraints directly by using a LayoutBuilder because it wasn't letting me put a LayoutBuilder inside of InstrincWidth...

But I'm pretty sure that the constraints are set to tight by the IntrinsicWidth widget.  If I remove IntrinsicWidth, the animation works.

Reply:
This and the comment in this review below should be irrelevant with the new approach in https://github.com/flutter/flutter/pull/49391/commits/42f976714798555c8e95e8708a499b6a1fd229fe.